packages/@react-aria/collections/src/CollectionBuilder.tsx

```diff
@@ -51,6 +52,12 @@ export function CollectionBuilder<C extends BaseCollection<object>>(props: Colle
   // This is fine. CollectionDocumentContext never changes after mounting.
   // eslint-disable-next-line react-hooks/rules-of-hooks
   let {collection, document} = useCollectionDocument(props.createCollection);
+
+  if (props.id) {
+    // @ts-expect-error
+    document.key = props.id;
+  }
```

Comment:
This came up in the a team discussion yesterday, but this approach doesn't work if there is an `id` provided to the collection component children (e.g. `<GridItem id="1" />`) unfortunately. If said `id` is provided, that becomes the `data-key` and said `id`s need only be unique per collection, not globally. This is already a problem for us in https://github.com/adobe/react-spectrum/blob/b37f2790b77543a447fb0dc1833b5d0cfa8d7da5/packages/%40react-aria/selection/src/useSelectableCollection.ts#L143
as you may have noticed, so we are trying to think of a better way to go about this (always prefix the data keys with a collection id? Create a weak map that ties the generated child ids to the collection and use that for lookup?).

I believe this is what you were trying to fix with the above change, lemme know if not otherwise

Reply:
Puh, it's been a while now 😆 IIRC, I also encountered this but concluded this case could be ignored, since at that point the item keys of the nested collection would also have to be manually set to a colliding value. If you want to make sure though, I would probably prefix them always.

Reply:
Said `node.key` is used in the generation of the whole collection key map setting up the next/prev relationship of each node in the collection, thus allowing us to do node look ups/return the key in selection or action/keyboard navigation etc. If the user provides their own `id`, we use that so that we can return the same `id` in `onSelectionChange`/`onAction` thus allowing the user to know which item is being acted upon.

Reply:
Thanks for circling back on this 👍 I do understand the purpose of the `node.key` in general and most of its associated functionality, so maybe I should clarify further.

What I was wondering about is why a developer is able to override the `node.key` composition (e.g. `react-aria-<key>`) by passing the `id` property to the leaf component. I'm asking because I don't understand why this was necessary in the first place.

As of now, when not customizing the `id` property, the internal `node.key` is passed to both action and selection callbacks. In this mode, the user simply has to strip the internal `react-aria` key prefix to access the item. Thus the only purpose i see this behavior currently solving is making the reverse mapping slightly more convenient. 

Am I missing additional side-effects this is addressing? If not, I'm unsure whether the existence of this feature is even warranted in the first place (other than backwards compatibility).

Reply:
We could've possibly retained the `react-aria` prefix and simply appended the user's provided `id` and handled stripping the prefix when calling `onSelectionChange` and/or `onAction`, I believe it just felt easier to replace the `node.key` at the time, unsure if there were any other factors. It also matched what we had in the old collections api in React Spectrum where we accepted `key`s from the user and used those in the collection map generation.